nixos/modules/services/networking/syncthing.nix

```diff
@@ -51,25 +53,91 @@ let
             --retry 1000 --retry-delay 1 --retry-all-errors \
             "$@"
     }
-
-    # query the old config
-    old_cfg=$(curl ${cfg.guiAddress}/rest/config)
-
-    # generate the new config by merging with the NixOS config options
-    new_cfg=$(printf '%s\n' "$old_cfg" | ${pkgs.jq}/bin/jq -c ${escapeShellArg ''. * ${builtins.toJSON cleanedConfig} * {
-        "devices": ('${escapeShellArg (builtins.toJSON devices)}'${optionalString (cfg.settings.devices == {} || ! cfg.overrideDevices) " + .devices"}),
-        "folders": ('${escapeShellArg (builtins.toJSON folders)}'${optionalString (cfg.settings.folders == {} || ! cfg.overrideFolders) " + .folders"})
-    }''})
-
-    # send the new config
-    curl -X PUT -d "$new_cfg" ${cfg.guiAddress}/rest/config
-
+  '' +
+
+  /* Syncthing's rest API for the folders and devices is almost identical.
+  Hence we iterate them using lib.pipe and generate shell commands for both at
+  the sime time. */
+  (lib.pipe ["folders" "devices"] [
+    # For the two above, create an attribute set with all the variables needed
+    # for writing the shell commands
+    (map (conf_type: {
+      inherit conf_type;
+      new_conf_IDs = lib.concatStringsSep " " {
+        devices = devicesIDs;
+        folders = foldersIDs;
+      }.${conf_type};
+      override = {
+        devices = cfg.overrideDevices;
+        folders = cfg.overrideFolders;
+      }.${conf_type};
+      conf = {
+        inherit devices folders;
+      }.${conf_type};
+      # The API's id naming is slightly different for devices and folders
+      GET_IdAttrName = {
+        devices = "deviceID";
+        folders = "id";
+      }.${conf_type};
+      # All URLs and curl commands are based on: https://docs.syncthing.net/rest/config.html
+      baseAddress = "${cfg.guiAddress}/rest/config/${conf_type}";
```

Comment:
not sure if building one attrset per device/folder would be more readable?
so I mean something like { devices = { new_conf_IDs = deviceIds; override = cfg.overrideDevices ...

Reply:
> not sure if building one attrset per device/folder would be more readable? so I mean something like { devices = { new_conf_IDs = deviceIds; override = cfg.overrideDevices ...

Yes I agree. I pushed something with even more usage of `lib.pipe` which perhaps will be better.